// pinus/dist/lib/common/remote/frontend/sessionRemote.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
function default_1(app) {
    return new SessionRemote(app);
}
exports.default = default_1;
class SessionRemote {
    constructor(app) {
        this.app = app;
    }
    bind(sid, uid) {
        return this.app.sessionService.abind(sid, uid);
    }
    unbind(sid, uid) {
        return this.app.sessionService.aunbind(sid, uid);
    }
    push(sid, key, value) {
        return this.app.sessionService.aimport(sid, key, value);
    }
    pushAll(sid, settings) {
        return this.app.sessionService.aimportAll(sid, settings);
    }
    /**
     * Get session informations with session id.
     *
     * @param  {String}   sid session id binded with the session
     * @param  {Function} cb(err, sinfo)  callback funtion, sinfo would be null if the session not exist.
     */
    getBackendSessionBySid(sid) {
        return __awaiter(this, void 0, void 0, function* () {
            let session = this.app.sessionService.get(sid);
            if (!session) {
                return;
            }
            return session.toFrontendSession().export();
        });
    }
    /**
     * Get all the session informations with the specified user id.
     *
     * @param  {String}   uid user id binded with the session
     * @param  {Function} cb(err, sinfo)  callback funtion, sinfo would be null if the session does not exist.
     */
    getBackendSessionsByUid(uid) {
        return __awaiter(this, void 0, void 0, function* () {
            let sessions = this.app.sessionService.getByUid(uid);
            if (!sessions) {
                return;
            }
            let res = [];
            for (let i = 0, l = sessions.length; i < l; i++) {
                res.push(sessions[i].toFrontendSession().export());
            }
            return res;
        });
    }
    /**
     * Kick a session by session id.
     *
     * @param  {Number}   sid session id
     * @param  {String}   reason  kick reason
     * @param  {Function} cb  callback function
     */
    kickBySid(sid, reason) {
        return this.app.sessionService.akickBySessionId(sid, reason);
    }
    /**
     * Kick sessions by user id.
     *
     * @param  {Number|String}   uid user id
     * @param  {String}          reason     kick reason
     * @param  {Function} cb     callback function
     */
    kickByUid(uid, reason) {
        return this.app.sessionService.kick(uid, reason);
    }
}
exports.SessionRemote = SessionRemote;
